File: src/domain/auth/guard.ts
/**
 * The authorization decision core (issue #46, SPEC §35 "Fine-grained
 * permissions / Organization isolation / Resource-level authorization").
 *
 * `can(principal, permission, resource?)` — DETERMINISTIC, DENY-BY-DEFAULT:
 *
 *   1. PERMISSION_UNKNOWN — the requested permission is not in the closed
 *      vocabulary (deny first: an unparseable request grants nothing);
 *   2. PRINCIPAL_SUSPENDED / PRINCIPAL_DEACTIVATED / PRINCIPAL_REVOKED —
 *      the principal record itself is not live;
 *   3. NO_GRANT — no active rule covers the permission (deny-by-default);
 *      NOT_IN_RESOURCE_SCOPE — the permission is held org-wide but not for
 *      THIS resource (grants may be resource-scoped; an org-wide grant —
 *      resourceId null — covers every resource).
 *
 * An ALLOW carries matched-rule EVIDENCE (the rule, its grant, its role) so
 * audit lines can say exactly which grant authorized the action.
 *
 * `authorize(...)` is the auditable boundary on top of `can`: it also
 * validates the optional session (idle/absolute expiry, revocation) and
 * wraps every denial as a DECISION VALUE paired with the `auth.accessDenied`
 * event — refusals are first-class facts (K2 precedent, SPEC §37), never
 * exceptions. Only a broken clock throws.
 */
import { DomainError, type Clock, type Uuid } from '../shared';
import { PERMISSIONS, type Role } from './roles';
import { isActiveGrant, type RoleGrant } from './assignments';
import type { ApiKey } from './apikeys';
import { sessionState } from './sessions';
import type { Session } from './sessions';
import type { User, UserStatus } from './user';
import { authEvent, type AccessDeniedPayload, type AuthEvent, type DenyReason } from './events';

/** Stable code carried by every audited denial decision value. */
export const AUTH_ACCESS_DENIED = 'AUTH_ACCESS_DENIED';

// --- the principal --------------------------------------------------------------------

/** One authority rule a principal holds (grant-derived or key scope). */
export interface PermissionRule {
  /** Concrete permission or `<resource>:*` role wildcard (expanded on match). */
  readonly rule: string;
  /** The role behind the rule (null for API-key scopes). */
  readonly roleId: Uuid | null;
  /** The grant behind the rule (null for API-key scopes). */
  readonly grantId: Uuid | null;
  /** Org-wide rule when null; scoped to exactly this resource otherwise. */
  readonly resourceId: Uuid | null;
}

/**
 * The resolved actor a decision runs against: the live status of the
 * principal record plus the rules its active grants/scopes confer.
 */
export interface Principal {
  readonly kind: 'user' | 'apiKey';
  readonly principalId: Uuid;
  readonly orgId: Uuid;
  readonly status: UserStatus | 'revoked';
  readonly rules: readonly PermissionRule[];
}

/**
 * Project a user + their grant facts + the org's role registry into a
 * Principal. Only ACTIVE grants contribute; grants whose role is missing
 * from the registry contribute NOTHING (deny-by-default — a dangling grant
 * can never confer authority). Role-level wildcards stay wildcards here and
 * are expanded at match time by `can`.
 */
export const userPrincipal = (
  user: User,
  assignments: readonly RoleGrant[],
  roles: readonly Role[],
): Principal => {
  const byId = new Map(roles.map((r) => [r.roleId, r]));
  const rules: PermissionRule[] = [];
  for (const grant of assignments) {
    if (grant.userId !== user.userId || !isActiveGrant(grant)) continue;
    const role = byId.get(grant.roleId);
    if (!role) continue;
    for (const rule of role.permissions) {
      rules.push({ rule, roleId: role.roleId, grantId: grant.grantId, resourceId: grant.resourceId });
    }
  }
  return {
    kind: 'user',
    principalId: user.userId,
    orgId: user.orgId,
    status: user.status,
    rules,
  };
};

/**
 * Project an API key into a Principal: its concrete scopes ARE its rules
 * (roleId/grantId null — evidence names the key itself), status 'revoked'
 * once the revocation fact landed.
 */
export const apiKeyPrincipal = (key: ApiKey): Principal => ({
  kind: 'apiKey',
  principalId: key.keyId,
  orgId: key.orgId,
  status: key.status === 'active' ? 'active' : 'revoked',
  rules: key.scopes.map((scope) => ({
    rule: scope,
    roleId: null,
    grantId: null,
    resourceId: null,
  })),
});

// --- the decision core ---------------------------------------------------------------------

export type CanDecision =
  | {
      readonly allowed: true;
      /** Matched-rule evidence: which rule, via which grant/role, in scope. */
      readonly via: PermissionRule;
    }
  | {
      readonly allowed: false;
      readonly reason: DenyReason;
      readonly detail: string;
    };

const STATUS_REASONS: Record<string, DenyReason> = {
  suspended: 'PRINCIPAL_SUSPENDED',
  deactivated: 'PRINCIPAL_DEACTIVATED',
  revoked: 'PRINCIPAL_REVOKED',
};

const wildcardCovers = (rule: string, permission: string): boolean =>
  rule.endsWith(':*') && permission.startsWith(rule.slice(0, -1));

/**
 * `can(principal, permission, resource?)` — the deterministic permission
 * matrix. Pure: no clock, no I/O; precedence pinned by tests.
 */
export function can(principal: Principal, permission: string, resource?: Uuid): CanDecision {
  // 1. Unknown permission — deny-by-default starts at the vocabulary.
  if (typeof permission !== 'string' || !isKnownPermission(permission)) {
    return {
      allowed: false,
      reason: 'PERMISSION_UNKNOWN',
      detail: `permission '${String(permission)}' is outside the closed vocabulary — nothing can be granted it`,
    };
  }
  // 2. The principal record itself must be live.
  if (principal.status !== 'active') {
    const reason = STATUS_REASONS[principal.status] ?? 'PRINCIPAL_DEACTIVATED';
    return {
      allowed: false,
      reason,
      detail: `principal ${principal.principalId} is ${principal.status} — no decision runs against an inactive principal`,
    };
  }
  // 3. Match rules: covering rule + (when a resource is named) scope.
  let coversPermission = false;
  for (const rule of principal.rules) {
    const covered = rule.rule === permission || wildcardCovers(rule.rule, permission);
    if (!covered) continue;
    coversPermission = true;
    if (resource === undefined || rule.resourceId === null || rule.resourceId === resource) {
      return { allowed: true, via: rule };
    }
  }
  return coversPermission
    ? {
        allowed: false,
        reason: 'NOT_IN_RESOURCE_SCOPE',
        detail: `permission '${permission}' is held but not for resource ${String(resource)} — grants are scoped, not guessed`,
      }
    : {
        allowed: false,
        reason: 'NO_GRANT',
        detail: `no active grant or scope covers '${permission}' — deny by default`,
      };
}

const KNOWN_PERMISSIONS: ReadonlySet<string> = new Set(PERMISSIONS);

const isKnownPermission = (permission: string): boolean => KNOWN_PERMISSIONS.has(permission);

// --- the auditable boundary --------------------------------------------------------------------

export interface AuthorizeArgs {
  /** The user record — null means "unknown principal" (still audited). */
  readonly user: User | null;
  readonly assignments: readonly RoleGrant[];
  readonly roles: readonly Role[];
  /** Present for session-authenticated requests; null for service calls. */
  readonly session: Session | null;
  readonly orgId: Uuid;
}

export type AuthorizeResult =
  | {
      readonly allowed: true;
      readonly decision: CanDecision & { readonly allowed: true };
      readonly principal: Principal;
    }
  | {
      readonly allowed: false;
      /** Stable code — always AUTH_ACCESS_DENIED (exported const). */
      readonly code: typeof AUTH_ACCESS_DENIED;
      readonly reason: DenyReason;
      readonly detail: string;
      /** auth.accessDenied — persist next to the denial decision (audit). */
      readonly event: AuthEvent<'auth.accessDenied', AccessDeniedPayload>;
    };

const SESSION_STATE_REASONS: Record<string, DenyReason> = {
  idleExpired: 'SESSION_IDLE_EXPIRED',
  absoluteExpired: 'SESSION_ABSOLUTE_EXPIRED',
  revoked: 'SESSION_REVOKED',
  ended: 'SESSION_ENDED',
  expired: 'SESSION_ABSOLUTE_EXPIRED',
};

/**
 * The auditable authorization boundary: session check → can() → wrap every
 * denial as a value + `auth.accessDenied` event (aggregate = the org; the
 * denial may concern an unknown principal). Pure: reads the arguments, never
 * mutates them; only a broken injected clock throws (AUTH_CLOCK_INVALID).
 */
export function authorize(
  args: AuthorizeArgs,
  permission: string,
  resource: Uuid | null,
  clock: Clock,
): AuthorizeResult {
  // Validate the clock up-front (also when the outcome is an early deny).
  const now = clock.now();
  if (!(now instanceof Date) || Number.isNaN(now.getTime())) {
    throw new DomainError('AUTH_CLOCK_INVALID', 'clock returned an invalid Date');
  }

  const deny = (reason: DenyReason, detail: string, principal: Principal | null): AuthorizeResult => ({
    allowed: false,
    code: AUTH_ACCESS_DENIED,
    reason,
    detail,
    event: authEvent(
      'auth.accessDenied',
      args.orgId,
      {
        orgId: args.orgId,
        principalId: principal === null ? null : principal.principalId,
        principalKind: principal === null ? 'unknown' : principal.kind,
        permission,
        resource,
        reason,
        detail,
        at: now.toISOString(),
      },
      clock,
    ),
  });

  // 0. Session gate (when the request rides a session).
  if (args.session) {
    const state = sessionState(args.session, clock);
    if (state !== 'active') {
      const reason = SESSION_STATE_REASONS[state] ?? 'SESSION_ENDED';
      return deny(
        reason,
        `session ${args.session.sessionId} is ${state} — expired or revoked sessions authorize nothing`,
        null,
      );
    }
  }

  // 1. Unknown principal.
  if (!args.user) {
    return deny('PRINCIPAL_UNKNOWN', 'no user record matches the presented identity', null);
  }

  // 2-4. The deterministic matrix.
  const principal = userPrincipal(args.user, args.assignments, args.roles);
  const decision = can(principal, permission, resource ?? undefined);
  if (decision.allowed) {
    return { allowed: true, decision, principal };
  }
  return deny(decision.reason, decision.detail, principal);
}

File: backend-go/internal/auth/guard.go
// Package auth is the Go port of the auth lane's authorization surface
// (src/domain/auth/{guard,roles,assignments}.ts) — the deterministic,
// deny-by-default permission matrix the kernel gates every route with.
package auth

import (
	"fmt"
	"sort"
	"strings"
)

// AUTHAccessDenied is the stable 403 code for authorization refusals.
const AUTHAccessDenied = "AUTH_ACCESS_DENIED"

// AdminManageUsersPermission is the role-administration permission —
// REQUIRED to grant or revoke roles (SPEC §35).
const AdminManageUsersPermission = "admin:manage-users"

// AUTHAttemptPermission is the marker used in audited payloads for
// authentication attempts themselves (not a vocabulary permission — the
// attempt is "be authenticated", authorization happens after).
const AUTHAttemptPermission = "auth:authenticate"

// NilOrg is the nil-org aggregate for denials that precede org
// identification (the TS middleware's NIL_ORG sentinel).
const NilOrg = "00000000-0000-4000-8000-000000000000"

// PERMISSIONS is the closed permission vocabulary (src/domain/auth/roles.ts).
// Adapters cannot invent a permission and have it mean anything.
var PERMISSIONS = []string{
	"receivables:read",
	"receivables:write",
	"payments:read",
	"payments:intake",
	"payments:refund",
	"collections:read",
	"collections:act",
	"adjustments:request",
	"adjustments:approve",
	"ledger:read",
	"ledger:post",
	"intelligence:read",
	"admin:manage-users",
	"policy:manage",
}

// RESOURCES is the resource half of the vocabulary (wildcards are legal only
// inside role definitions, exactly one per known resource).
var RESOURCES = []string{
	"receivables", "payments", "collections", "adjustments",
	"ledger", "intelligence", "admin", "policy",
}

var permissionSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(PERMISSIONS))
	for _, p := range PERMISSIONS {
		set[p] = struct{}{}
	}
	return set
}()

var resourceSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(RESOURCES))
	for _, r := range RESOURCES {
		set[r] = struct{}{}
	}
	return set
}()

// IsKnownPermission reports whether the permission is in the closed
// vocabulary — deny-by-default starts here.
func IsKnownPermission(permission string) bool {
	_, ok := permissionSet[permission]
	return ok
}

// IsRoleWildcard reports whether raw is exactly `<known-resource>:*` — a
// role-level wildcard (legal ONLY inside role definitions, never per grant
// or key scope).
func IsRoleWildcard(raw string) bool {
	idx := strings.Index(raw, ":")
	if idx <= 0 || !strings.HasSuffix(raw, ":*") {
		return false
	}
	_, ok := resourceSet[raw[:idx]]
	return ok && raw == raw[:idx]+":*"
}

func isPermissionShape(raw string) bool {
	// /^[a-z][a-z0-9]*:[a-z][a-zA-Z0-9-]*$/
	parts := strings.SplitN(raw, ":", 2)
	if len(parts) != 2 {
		return false
	}
	if !lowerIdent(parts[0]) || parts[1] == "" {
		return false
	}
	for i, c := range parts[1] {
		alnum := c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9'
		if !(alnum || c == '-' && i > 0) {
			return false
		}
	}
	return parts[1][0] >= 'a' && parts[1][0] <= 'z'
}

func lowerIdent(s string) bool {
	if s == "" || s[0] < 'a' || s[0] > 'z' {
		return false
	}
	for _, c := range s {
		if !(c >= 'a' && c <= 'z' || c >= '0' && c <= '9') {
			return false
		}
	}
	return true
}

// AssertPermission validates a CONCRETE permission (the only kind allowed in
// grants and API-key scopes). Refusals (stable codes):
//   - AUTH_PERMISSION_WILDCARD_FORBIDDEN — a wildcard where only concrete
//     permissions are allowed;
//   - AUTH_PERMISSION_MALFORMED — not `<resource>:<action>` shaped;
//   - AUTH_PERMISSION_UNKNOWN — well-formed but outside the vocabulary.
func AssertPermission(raw string) (string, error) {
	if IsRoleWildcard(raw) {
		return "", fmt.Errorf("AUTH_PERMISSION_WILDCARD_FORBIDDEN: wildcard '%s' is only allowed inside role definitions, never per grant or key scope", raw)
	}
	if !isPermissionShape(raw) {
		return "", fmt.Errorf("AUTH_PERMISSION_MALFORMED: permission '%s' is not a '<resource>:<action>' string", raw)
	}
	if !IsKnownPermission(raw) {
		return "", fmt.Errorf("AUTH_PERMISSION_UNKNOWN: permission '%s' is outside the closed vocabulary", raw)
	}
	return raw, nil
}

// PermissionRule is one authority rule a principal holds (grant-derived or
// key scope).
type PermissionRule struct {
	// Rule is a concrete permission or a `<resource>:*` role wildcard
	// (expanded on match).
	Rule string
	// RoleID / GrantID are the evidence trail (empty for key scopes).
	RoleID  string
	GrantID string
	// ResourceID is empty for org-wide rules; scoped otherwise.
	ResourceID string
}

// Principal is the resolved actor a decision runs against.
type Principal struct {
	Kind        string // "user" | "apiKey"
	PrincipalID string
	OrgID       string
	Status      string // active | suspended | deactivated | revoked
	Rules       []PermissionRule
}

// CanDecision is the deterministic decision value: either allowed with
// matched-rule evidence, or refused with a stable reason + detail.
type CanDecision struct {
	Allowed bool
	// Via is the matched rule (evidence for audit lines) when allowed.
	Via PermissionRule
	// Reason + Detail carry the refusal (DenyReason + human detail).
	Reason string
	Detail string
}

var statusReasons = map[string]string{
	"suspended":   "PRINCIPAL_SUSPENDED",
	"deactivated": "PRINCIPAL_DEACTIVATED",
	"revoked":     "PRINCIPAL_REVOKED",
}

func wildcardCovers(rule, permission string) bool {
	return strings.HasSuffix(rule, ":*") && strings.HasPrefix(permission, rule[:len(rule)-1])
}

// Can is the deterministic permission matrix (the port of guard.ts `can`):
//
//  1. PERMISSION_UNKNOWN — the requested permission is outside the vocabulary;
//  2. PRINCIPAL_SUSPENDED / _DEACTIVATED / _REVOKED — the principal record
//     itself is not live;
//  3. NO_GRANT — no active rule covers the permission (deny-by-default).
//
// The mounted /v1 surface authorizes org-wide (no resource parameter), so a
// covering rule — org-wide or resource-scoped — allows; scope-refusal edges
// (NOT_IN_RESOURCE_SCOPE) only exist behind the guard's resource argument,
// which this kernel never passes. Pure: no clock, no I/O.
func Can(p Principal, permission string) CanDecision {
	if !IsKnownPermission(permission) {
		return CanDecision{
			Allowed: false,
			Reason:  "PERMISSION_UNKNOWN",
			Detail:  fmt.Sprintf("permission '%s' is outside the closed vocabulary — nothing can be granted it", permission),
		}
	}
	if p.Status != "active" {
		reason := statusReasons[p.Status]
		if reason == "" {
			reason = "PRINCIPAL_DEACTIVATED"
		}
		return CanDecision{
			Allowed: false,
			Reason:  reason,
			Detail:  fmt.Sprintf("principal %s is %s — no decision runs against an inactive principal", p.PrincipalID, p.Status),
		}
	}
	for _, rule := range p.Rules {
		if rule.Rule == permission || wildcardCovers(rule.Rule, permission) {
			return CanDecision{Allowed: true, Via: rule}
		}
	}
	return CanDecision{
		Allowed: false,
		Reason:  "NO_GRANT",
		Detail:  fmt.Sprintf("no active grant or scope covers '%s' — deny by default", permission),
	}
}

// EffectivePermissions expands a principal's rule set into the concrete
// permission set used by the escalation guard (role wildcards expand to
// every vocabulary permission on their resource). Sorted, deduped.
func EffectivePermissions(rules []PermissionRule) []string {
	set := make(map[string]struct{})
	for _, rule := range rules {
		if strings.HasSuffix(rule.Rule, ":*") {
			prefix := rule.Rule[:len(rule.Rule)-1]
			for _, p := range PERMISSIONS {
				if strings.HasPrefix(p, prefix) {
					set[p] = struct{}{}
				}
			}
			continue
		}
		set[rule.Rule] = struct{}{}
	}
	out := make([]string, 0, len(set))
	for p := range set {
		out = append(out, p)
	}
	sort.Strings(out)
	return out
}

// MissingForRole returns the sorted concrete permissions (or wildcard rules)
// of a role definition that the granter's effective set does NOT cover — the
// escalation guard's "a granter cannot confer authority they do not hold".
func MissingForRole(roleRules, granterPermissions []string) []string {
	var missing []string
	for _, rule := range roleRules {
		if strings.HasSuffix(rule, ":*") {
			prefix := rule[:len(rule)-1]
			covers := false
			for _, p := range granterPermissions {
				if strings.HasPrefix(p, prefix) {
					covers = true
					break
				}
			}
			if !covers {
				missing = append(missing, rule)
			}
			continue
		}
		held := false
		for _, p := range granterPermissions {
			if p == rule {
				held = true
				break
			}
		}
		if !held {
			missing = append(missing, rule)
		}
	}
	sort.Strings(missing)
	return missing
}
